File: erettsegi_emelt/2021_okt/java/Sudoku.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Sudoku {

    public static void main(String[] args) throws IOException {
        System.out.println("1. Feladat: Írd be 1 bemeneti fájl nevét, 1 sor és 1 oszlop számát!");

        var input = new Scanner(System.in);
        var inputFileName = input.nextLine();
        var inputRowIndex = input.nextInt() - 1;
        var inputColumnIndex = input.nextInt() - 1;

        input.close();

        var lines = Files.readAllLines(Path.of(inputFileName));

        var numbersPerLine = new int[lines.size()][];
        for(var j = 0; j < lines.size(); j++) {
            var line = lines.get(j);
            var split = line.split(" ");
            var values = new int[split.length];

            for(var i = 0; i < split.length; ++i) {
                values[i] = Integer.parseInt(split[i]);
            }

            numbersPerLine[j] = values;
        }

        var gameState = Arrays.copyOfRange(numbersPerLine, 0, 9);

        var valueAtInputSlot = gameState[inputRowIndex][inputColumnIndex];
        var tableNumber = 3 * (inputRowIndex / 3) + (inputColumnIndex / 3) + 1;
        var valueToPrint = valueAtInputSlot == 0 ? "Az adott helyet még nem töltötték ki!" : ("Adott helyen szereplő szám: " + valueAtInputSlot);

        System.out.println("3. Feladat: " + valueToPrint + ", résztáblázat száma: " + tableNumber);

        var unfilledSlotCount = 0;
        for(var row : gameState) {
            for(var value : row) {
                if(value == 0) {
                    ++unfilledSlotCount;
                }
            }
        }

        System.out.printf("4. Feladat: Üres helyek aránya: %.1f%%\n", unfilledSlotCount / 81F * 100F);
        System.out.println("5. Feladat:");

        for(var i = 9; i < numbersPerLine.length; ++i) {
            var k = numbersPerLine[i];

            System.out.println("Sor: " + k[1] + ", oszlop: " + k[2] + ", érték: " + k[0] + ": " + getStepAttemptResultMessage(k[0], k[1] - 1, k[2] - 1, gameState));
        }
    }

    private static String getStepAttemptResultMessage(int value, int rowIndex, int columnIndex, int[][] gameState) {
        if(gameState[rowIndex][columnIndex] != 0) return "A helyet már kitöltötték";

        for(var values : gameState[rowIndex]) {
            if(values == value) {
                return "Az adott sorban már szerepel a szám";
            }
        }

        for(var row : gameState) {
            if(row[columnIndex] == value) {
                return "Az adott oszlopban már szerepel a szám";
            }
        }

        var beginRow = (rowIndex / 3) * 3;
        var endRow = beginRow + 3;
        var beginColumn = (columnIndex / 3) * 3;
        var endColumn = beginColumn + 3;

        for(var i = beginRow; i < endRow; ++i) {
            for(var k = beginColumn; k < endColumn; ++k) {
                if(gameState[i][k] == value) {
                    return "Az adott résztáblában már szerepel a szám";
                }
            }
        }

        return "A lépés megtehető";
    }
}

File: erettsegi_emelt/2021_okt/python/Sudoku.py
def get_step_attempt_result_message(value: int, row_index: int, column_index: int, game_state: list[list[int]]):
    if game_state[row_index][column_index] != 0: return "A helyet már kitöltötték"
    if any(k for k in game_state[row_index] if k == value): return "Az adott sorban már szerepel a szám"
    if any(k for k in game_state if k[column_index] == value): return "Az adott oszlopban már szerepel a szám"

    begin_row = (row_index // 3) * 3
    end_row = begin_row + 3
    begin_column = (column_index // 3) * 3
    end_column = begin_column + 3

    for numbers_for_line in game_state[begin_row : end_row]:
        for num in numbers_for_line[begin_column : end_column]:
            if num == value:
                return "Az adott résztáblában már szerepel a szám"

    return 'A lépés megtehető'


print('1. Feladat: Írd be 1 bemeneti fájl nevét, 1 sor és 1 oszlop számát!')

input_file_name = input()
input_row_index = int(input()) - 1
input_column_index = int(input()) - 1

with open(input_file_name) as input_file:
    lines = input_file.readlines()

    parse_line = lambda split_line: [ int(n) for n in split_line ]
    numbers_per_line = [ parse_line(k.split(' ')) for k in lines ]

game_state = numbers_per_line[0 : 9]

value_at_input_slot = game_state[input_row_index][input_column_index]
table_number = 3 * (input_row_index // 3) + (input_column_index // 3) + 1
value_to_print = 'Az adott helyet még nem töltötték ki!' if value_at_input_slot == 0 else (f'Adott helyen szereplő szám: {value_at_input_slot}')

print(f'3. Feladat: {value_to_print}, résztáblázat száma: {table_number}')

unfilled_slot_count = 0
for numbers_in_line in game_state:
    for num in numbers_in_line:
        if num == 0:
            unfilled_slot_count += 1

print(f'4. Feladat: Üres helyek aránya: {(unfilled_slot_count / 81 * 100):.1f}')
print('5. Feladat:')

for numbers_in_line in numbers_per_line[9 :]:
    print(f'Sor: {numbers_in_line[1]}, oszlop: {numbers_in_line[2]}, érték: {numbers_in_line[0]}: {get_step_attempt_result_message(numbers_in_line[0], numbers_in_line[1] - 1, numbers_in_line[2] - 1, game_state)}')